pingcap/tiup
# playground: implement `mode` option for painter algorithm populating options

### What problem does this PR solve?

This PR closes #1269 

It introduces a `mode` option to switch default config set among `tidb` or `tikv-slim` and allow user populate more options based on the default set.

### What is changed and how it works?

Now, we directly communicate with `FlagSet` and use `Changed` to see if user explicit sets a config flag.

And I implement the painter algorithm that

1. layer 0: draw the `version` config
2. layer 2: draw the default set based on `mode` option
3. layer 3: draw the reset options configured by user

each layer overwrite options set in early layer.

### Check List

Tests

 - Manual test (add detailed scripts or steps below)

I run

- `tiup playground`
- `tiup playground --mode tidb`
- `tiup playground --mode tikv-slim`
- `tiup playground --mode unknown`
- `tiup playground --db 0 --tiflash 0`
- `tiup playground --mode tikv-slim --db 1 --tiflash 1`

and all run as expected.

I'm willing to add some automatic test but help wanted. Also I think we can add tests after this pr merged instead of be too mean to a fruitful pull request.

Side effects

 - Increased code complexity

with increasing flexibility.

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

Yes, we need to update the documentation. But where?

```release-note
Implement `mode` option for painter algorithm populating options. `tidb` as the original defaults and `tikv-slim` for only pd, tikv, and monitor instances.
```
